File: apps/worker/services/transcodeVideo.ts
import {
  GetObjectCommand,
  PutObjectCommand,
  S3Client,
} from "@aws-sdk/client-s3";

import { mkdir, readdir, rm } from "node:fs/promises";
import { createWriteStream } from "node:fs";
import { pipeline } from "node:stream/promises";
import { Readable } from "node:stream";
import { basename, join, posix, relative, sep } from "node:path";

const s3 = new S3Client({
  region: process.env.AWS_REGION,
  credentials: {
    accessKeyId: process.env.AWS_ACCESS_KEY_ID!,
    secretAccessKey: process.env.AWS_SECRET_ACCESS_KEY!,
  },
});

const bucketName = process.env.S3_BUCKET_NAME!;

/*
Main transcoding function.

Example input:

await transcodeVideo({
  sourceKey: "uploads/user-123/holiday.mp4",
});

Example return value:

{
  sourceKey: "uploads/user-123/holiday.mp4",
  previewPrefix: "previews/uploads/user-123/holiday",
  masterPlaylistKey:
    "previews/uploads/user-123/holiday/master.m3u8"
}
*/

export async function transcodeVideo({ sourceKey }: { sourceKey: string }) {
  if (!sourceKey) {
    throw new Error("sourceKey is required");
  }

  const workDir = join(import.meta.dir, "..", "tmp", crypto.randomUUID());
  /*
  uploads/user-123/holiday.mp4
  getFileNameFromKey(sourceKey) returns:
  holiday.mp4
  Example inputPath:
  /app/tmp/random-job-id/holiday.mp4
  */

  const inputPath = join(workDir, getFileNameFromKey(sourceKey));
  /*
  Local directory where FFmpeg will generate all HLS files.
  Example:
  /app/tmp/random-job-id/hls
  */
  const hlsOutputDir = join(workDir, "hls");

  /*
  Creates the base S3 path for this video's generated preview files.

  Example input:

  uploads/user-123/holiday.mp4

  createPreviewPrefix() returns:

  previews/uploads/user-123/holiday

  All HLS files will be uploaded under this prefix.
  */

  const previewPrefix = createPreviewPrefix(sourceKey);

  /*
  Creates the exact S3 key for the master HLS playlist.

  previewPrefix:

  previews/uploads/user-123/holiday

  Result:

  previews/uploads/user-123/holiday/master.m3u8

  This is the main file that the frontend video player loads.
  */

  const masterPlaylistKey = posix.join(previewPrefix, "master.m3u8");

  await mkdir(hlsOutputDir, { recursive: true });

  try {
    console.log("Downloading original:", sourceKey);
    //  /app/tmp/random-job-id/holiday.mp4
    await downloadFromS3(sourceKey, inputPath);

    console.log("creating hls preview");
    await createHlsPreview(inputPath, hlsOutputDir);

    console.log("uploading HLS preview:", previewPrefix);
    // previews/uploads/user-123/holiday
    await uploadDirectoryToS3(hlsOutputDir, previewPrefix);

    console.log("HLS preview created", masterPlaylistKey);

    return {
      sourceKey,
      previewPrefix,
      masterPlaylistKey,
    };
  } finally {
    await rm(workDir, {
      recursive: true,
      force: true,
    });
  }
}

async function downloadFromS3(key: string, destinationPath: string) {
  const result = await s3.send(
    new GetObjectCommand({
      Bucket: bucketName,
      Key: key,
    }),
  );
  if (!result.Body) {
    throw new Error("S3 object body is empty");
  }
  await pipeline(result.Body as Readable, createWriteStream(destinationPath));
}

async function createHlsPreview(inputPath: string, outputDir: string) {
  const ffmpeg = Bun.spawn({
    cmd: [
      "ffmpeg",
      "-y",
      "-i",
      inputPath,

      "-filter_complex",
      "[0:v]split=4[v360][v480][v720][v1080];[v360]scale=-2:360[v360out];[v480]scale=-2:480[v480out];[v720]scale=-2:720[v720out];[v1080]scale=-2:1080[v1080out]",

      "-map",
      "[v360out]",
      "-map",
      "0:a?",
      "-c:v:0",
      "libx264",
      "-b:v:0",
      "800k",
      "-maxrate:v:0",
      "900k",
      "-bufsize:v:0",
      "1200k",
      "-c:a:0",
      "aac",
      "-b:a:0",
      "96k",

      "-map",
      "[v480out]",
      "-map",
      "0:a?",
      "-c:v:1",
      "libx264",
      "-b:v:1",
      "1400k",
      "-maxrate:v:1",
      "1600k",
      "-bufsize:v:1",
      "2200k",
      "-c:a:1",
      "aac",
      "-b:a:1",
      "128k",

      "-map",
      "[v720out]",
      "-map",
      "0:a?",
      "-c:v:1",
      "libx264",
      "-b:v:1",
      "2500k",
      "-maxrate:v:1",
      "2800k",
      "-bufsize:v:1",
      "4000k",
      "-c:a:1",
      "aac",
      "-b:a:1",
      "128k",

      "-map",
      "[v1080out]",
      "-map",
      "0:a?",
      "-c:v:2",
      "libx264",
      "-b:v:2",
      "5000k",
      "-maxrate:v:2",
      "5500k",
      "-bufsize:v:2",
      "8000k",
      "-c:a:2",
      "aac",
      "-b:a:2",
      "192k",

      "-f",
      "hls",
      "-hls_time",
      "6",
      "-hls_playlist_type",
      "vod",
      "-hls_segment_filename",
      join(outputDir, "stream_%v", "segment_%03d.ts"),
      "-master_pl_name",
      "master.m3u8",
      "-var_stream_map",
      "v:0,a:0,name:360p v:1,a:1,name:480p v:2,a:2,name:720p v:3,a:3,name:1080p",
      join(outputDir, "stream_%v", "index.m3u8"),
    ],
    stdout: "pipe",
    stderr: "pipe",
  });
  const exitCode = await ffmpeg.exited;
  if (exitCode !== 0) {
    const errorOutput = await new Response(ffmpeg.stderr).text();

    throw new Error(`FFmpeg failed with exit code ${exitCode}\n${errorOutput}`);
  }
}

async function uploadFileToS3WithRetry(filePath: string, key: string) {
  const maxAttempts = 3;

  for (let attempt = 1; attempt <= maxAttempts; attempt++) {
    try {
      const file = Bun.file(filePath);
      const arrayBuffer = await file.arrayBuffer();
      const bytes = new Uint8Array(arrayBuffer);

      await s3.send(
        new PutObjectCommand({
          Bucket: bucketName,
          Key: key,
          Body: bytes,
          ContentType: getContentType(filePath),
          ContentLength: bytes.length,
        }),
      );

      return;
    } catch (error) {
      console.error(`Upload failed for ${key}, attempt ${attempt}`, error);

      if (attempt === maxAttempts) {
        throw error;
      }

      await new Promise((resolve) => setTimeout(resolve, attempt * 2000));
    }
  }
}

async function uploadDirectoryToS3(localDir: string, s3Prefix: string) {
  const files = await getFilesRecursively(localDir);

  for (const filePath of files) {
    const relativePath = relative(localDir, filePath).split(sep).join("/");
    const key = posix.join(s3Prefix, relativePath);

    await uploadFileToS3WithRetry(filePath, key);

    console.log("uploaded Key", key);
  }
}

async function getFilesRecursively(dir: string): Promise<string[]> {
  const entries = await readdir(dir, {
    withFileTypes: true,
  });

  const files = await Promise.all(
    entries.map((entry) => {
      const fullPath = join(dir, entry.name);
      if (entry.isDirectory()) {
        return getFilesRecursively(fullPath);
      }
      return [fullPath];
    }),
  );
  return files.flat();
}

function createPreviewPrefix(sourceKey: string) {
  const parsed = posix.parse(sourceKey);

  return posix.join("previews", parsed.dir, parsed.name);
}

function getFileNameFromKey(key: string) {
  return basename(key) || "source-video";
}

function getContentType(filePath: string) {
  if (filePath.endsWith(".m3u8")) {
    return "application/vnd.apple.mpegurl";
  }

  if (filePath.endsWith(".ts")) {
    return "video/mp2t";
  }

  return "application/octet-stream";
}

File: apps/go-api/internal/tasks/transcodeVideo.go
package tasks

import (
	"context"
	"database/sql"
	"fmt"
	"io"
	"log"
	"os"
	"os/exec"
	"path"
	"path/filepath"
	"strings"
	"time"

	"github.com/aws/aws-sdk-go-v2/aws"
	"github.com/aws/aws-sdk-go-v2/service/s3"
	"github.com/google/uuid"
)

type Processor struct {
	S3     *s3.Client
	Bucket string
	DB     *sql.DB
}

/*
S3:
uploads/videos/movie.mp4
        │
        │ downloadFromS3()
        ▼
/tmp/uploadrelay/xyz/movie.mp4
        │
        │ createHLSPreview()
        ▼
hls/
├── master.m3u8
├── stream_360p/
│   ├── index.m3u8
│   ├── segment_000.ts
│   ├── segment_001.ts
│   └── ...
│
└── stream_720p/
    ├── index.m3u8
    ├── segment_000.ts
    ├── segment_001.ts
    └── ...
        │
        │ uploadDirToS3()
        ▼
S3:
previews/...
*/

func (p *Processor) TranscodeVideo(ctx context.Context, sourceKey string) error {
	if sourceKey == "" {
		return fmt.Errorf("source_key is required")
	}

	workDir := filepath.Join(os.TempDir(), "uploadRelay", uuid.NewString())
	inputPath := filepath.Join(workDir, fileNameFromKey(sourceKey))
	hlsDir := filepath.Join(workDir, "hls")
	previewPrefix := createPreviewPrefix(sourceKey)
	masterKey := path.Join(previewPrefix, "master.m3u8")

	if err := os.MkdirAll(filepath.Join(hlsDir, "stream_360p"), 0o755); err != nil {
		return err
	}
	if err := os.MkdirAll(filepath.Join(hlsDir, "stream_720p"), 0o755); err != nil {
		return err
	}
	defer os.RemoveAll(workDir)

	log.Printf("downloading %s", sourceKey)

	if err := p.downloadFromS3(ctx, sourceKey, inputPath); err != nil {
		return err
	}

	log.Println("running ffmpeg")

	if err := createHLSPreview(ctx, inputPath, hlsDir); err != nil {
		return err
	}

	log.Printf("uploading HLS to %s", previewPrefix)

	if err := p.uploadDirToS3(ctx, hlsDir, previewPrefix); err != nil {
		return err
	}

	log.Printf("done: %s", masterKey)
	log.Println("completed the task")
	return nil
}

func (p *Processor) downloadFromS3(ctx context.Context, key, dest string) error {
	out, err := p.S3.GetObject(ctx, &s3.GetObjectInput{
		Key:    aws.String(key),
		Bucket: aws.String(p.Bucket),
	})
	if err != nil {
		return fmt.Errorf("get object: %w", err)
	}
	defer out.Body.Close()

	f, err := os.Create(dest)
	if err != nil {
		return err
	}
	defer f.Close()

	if _, err := io.Copy(f, out.Body); err != nil {
		return fmt.Errorf("write file: %w", err)
	}
	return nil
}

func createHLSPreview(ctx context.Context, inputPath, outputDir string) error {
	// Forward slashes: ffmpeg %v substitution breaks on Windows backslashes.
	segmentPattern := filepath.ToSlash(
		filepath.Join("stream_%v", "segment_%03d.ts"),
	)
	playlistPattern := filepath.ToSlash(
		filepath.Join("stream_%v", "index.m3u8"),
	)
	cmd := exec.CommandContext(ctx, "ffmpeg",
		"-y", "-i", inputPath,
		"-filter_complex", "[0:v]split=2[v360][v720];[v360]scale=-2:360[v360out];[v720]scale=-2:720[v720out]",
		"-map", "[v360out]", "-map", "0:a?",
		"-c:v:0", "libx264", "-preset:v:0", "veryfast", "-crf:v:0", "24",
		"-b:v:0", "700k", "-maxrate:v:0", "850k", "-bufsize:v:0", "1200k",
		"-c:a:0", "aac", "-b:a:0", "96k",
		"-map", "[v720out]", "-map", "0:a?",
		"-c:v:1", "libx264", "-preset:v:1", "veryfast", "-crf:v:1", "24",
		"-b:v:1", "2200k", "-maxrate:v:1", "2600k", "-bufsize:v:1", "3600k",
		"-c:a:1", "aac", "-b:a:1", "128k",
		"-f", "hls",
		"-hls_time", "8",
		"-hls_playlist_type", "vod",
		"-hls_segment_filename", segmentPattern,
		"-master_pl_name", "master.m3u8",
		"-var_stream_map", "v:0,a:0,name:360p v:1,a:1,name:720p",
		playlistPattern,
	)
	cmd.Dir = outputDir
	out, err := cmd.CombinedOutput()
	if err != nil {
		return fmt.Errorf("ffmpeg failed: %w\n%s", err, out)
	}
	return nil
}

/*
These files are independent. Uploading segment_001.ts does not depend on uploading segment_000.ts. So you can upload several at the same time.
But don’t start unlimited goroutines. Use a small concurrency limit, like 4.
Conceptually:
upload 4 files at once
when one finishes, start another
if any upload fails, return error
*/

func (p *Processor) uploadDirToS3(ctx context.Context, localDir, s3Prefix string) error {
	return filepath.Walk(localDir, func(fp string, info os.FileInfo, err error) error {
		if err != nil {
			return err
		}
		if info.IsDir() {
			return nil
		}
		rel, err := filepath.Rel(localDir, fp)
		if err != nil {
			return err
		}
		key := path.Join(s3Prefix, filepath.ToSlash(rel))
		return p.uploadFileWithRetry(ctx, fp, key)
	})
}

func (p *Processor) uploadFileWithRetry(ctx context.Context, filePath, key string) error {
	var last error
	for attempt := 1; attempt <= 3; attempt++ {
		last = p.uploadFile(ctx, filePath, key)
		if last == nil {
			log.Printf("uploaded %s", key)
			return nil
		}
		log.Printf("upload failed %s attempt %d: %v", key, attempt, last)
		time.Sleep(time.Duration(attempt) * 2 * time.Second)
	}
	return last
}

func (p *Processor) uploadFile(ctx context.Context, filePath, key string) error {
	f, err := os.Open(filePath)
	if err != nil {
		return err
	}
	defer f.Close()
	stat, err := f.Stat()
	if err != nil {
		return err
	}
	_, err = p.S3.PutObject(ctx, &s3.PutObjectInput{
		Bucket:        aws.String(p.Bucket),
		Key:           aws.String(key),
		Body:          f,
		ContentType:   aws.String(contentType(filePath)),
		ContentLength: aws.Int64(stat.Size()),
	})
	return err
}

func createPreviewPrefix(sourceKey string) string {
	dir := path.Dir(sourceKey)
	name := strings.TrimSuffix(path.Base(sourceKey), path.Ext(sourceKey))
	return path.Join("previews", dir, name)
}

func fileNameFromKey(key string) string {
	base := path.Base(key)
	if base == "." || base == "/" {
		return "source-video"
	}
	return base
}

func contentType(filePath string) string {
	switch strings.ToLower(filepath.Ext(filePath)) {
	case ".m3u8":
		return "application/vnd.apple.mpegurl"
	case ".ts":
		return "video/mp2t"
	default:
		return "application/octet-stream"
	}
}
